Initial migration of type system examples (#3340)
dart-lang/site-www: src/_guides/language/type-system.md

[Try it in DartPad]({{site.dartpad}}/4a2a9bc2242042ba5338533d091213c0).
{% endcomment %}

What about going in the other direction? 
Can you assign an `Animal` list to a `List<Cat>`?

{:.passes-sa}
<?code-excerpt "lib/strong_analysis.dart (generic-type-assignment-Animal)" replace="/Animal/[!$&!]/g"?>
{% prettify dart tag=pre+code %}
List<Cat> myCats = <[!Animal!]>[];
{% endprettify %}

> This assignment does not pass static analysis 
> as it creates an implicit downcast, 
> which are disallowed from non-`dynamic` types such as `Animal`.

Reviewer: ```suggestion
This assignment doesn't pass static analysis 
because it creates an implicit downcast, 
which is disallowed from non-`dynamic` types such as `Animal`.
```
https://developers.google.com/style/contractions
https://developers.google.com/style/word-list#as

{{site.alert.version-note}}
  In packages that use a [language version][] before 2.12
  (when support for [null safety][] was introduced),
  code can implicitly downcast from these non-`dynamic` types.

>   You can disallow these non-`dynamic` downcasts
>   by specifying `implicit-casts: false` in the [analysis options file.][analysis]

Reviewer: ```suggestion
  You can disallow non-`dynamic` downcasts in a pre-2.12 project
  by specifying `implicit-casts: false` in the [analysis options file.][analysis]
```

{{site.alert.end}}

> For this code to pass static analysis, 
> you will instead need to use an explicit cast, 
> which may fail at runtime.

Reviewer: ```suggestion
To make this code pass static analysis, 
use an explicit cast, 
which might fail at runtime.
```

Tweaks to make it more direct and use [present tense](https://developers.google.com/style/tense), plus avoid the word [_may_](https://developers.google.com/style/word-list#may).

The following resources have further information on sound Dart:

* [Fixing common type problems](/guides/language/sound-problems) - 
  Errors you may encounter when writing sound Dart code, and how to fix them.
* [Fixing type promotion failures](/tools/non-promotion-reasons) - 
  Understand and learn how to fix type promotion errors.
* [Sound null safety](/null-safety) - 
  Learn about writing code with sound null safety enabled.
* [Customizing static analysis][analysis] - 
  How to set up and customize the analyzer and linter
  using an analysis options file.

> * [Dart 2](/dart-2) - 

Reviewer: ```suggestion
* [Dart 2 migration guide](/dart-2) - 
```
